src/index.ts

```diff
@@ -1,27 +1,34 @@
 import { compileVM } from 'whynot';
 import { compileRegExp } from './compiler';
-import { parse } from './parser';
+import { generateParser } from './parser';
 
 function toCodePoints(str: string): number[] {
 	return [...str].map(c => c.codePointAt(0)!);
 }
 
 export type MatchFn = (str: string) => boolean;
 
+export type Options = {
+	language: 'xsd' | 'xpath';
+};
+
 /**
  * @param pattern Pattern to compile
+ * @param options Additional options for the compiler
  *
  * @return a matcher function, or null if the pattern uses unsupported features
  */
-export function compile(pattern: string): MatchFn {
-	const ast = parse(pattern);
+export function compile(pattern: string, options: Options = { language: 'xsd' }): MatchFn {
+	let ast = generateParser(options)(pattern);
 
 	const vm = compileVM<number>(assembler => {
 		compileRegExp(assembler, ast);
 		assembler.accept();
 	});
 
 	return function match(str: string): boolean {
-		return vm.execute(toCodePoints(str)).success;
+		const codepoints =
+			options.language === 'xpath' ? [-1, ...toCodePoints(str), -2] : toCodePoints(str);
```

Comment:
I think rather than having special tokens, the start and end of the input string are better handled by a special parser checking for the corresponding offset (and not consuming any input).

Reply:
You can use prsc's `end` parser for $. Let's add a `start` parser over there and then use those two here.

Reply:
Ignore that, I was thinking about the wrong parser.